microceph/ceph/osd.go

```diff
@@ -124,24 +124,48 @@ func AddOSD(s *state.State, path string, wipe bool) error {
 		return fmt.Errorf("Failed to find next OSD number: %w", err)
 	}
 
+	// Record the disk.
+	err = s.Database.Transaction(s.Context, func(ctx context.Context, tx *sql.Tx) error {
+		_, err := database.CreateDisk(ctx, tx, database.Disk{Member: s.Name(), Path: path, OSD: int(nr)})
+		if err != nil {
+			return fmt.Errorf("Failed to record disk: %w", err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// if we fail later, make sure we free up the record
+	cleanup := func() {
+		s.Database.Transaction(s.Context, func(ctx context.Context, tx *sql.Tx) error {
+			database.DeleteDisk(ctx, tx, s.Name(), path)
+			return nil
+		})
+	}
+
 	// Create directory.
 	dataPath := filepath.Join(os.Getenv("SNAP_COMMON"), "data")
 	osdDataPath := filepath.Join(dataPath, "osd", fmt.Sprintf("ceph-%d", nr))
 
 	err = os.MkdirAll(osdDataPath, 0700)
 	if err != nil {
+		cleanup()
```

Comment:
why not `defer`ing the cleanup callback? This way is a bit too error prone imo.

Reply:
yeah, I had the same thought; except defer is the thing that always runs at the end, AIUI

Reply:
there does also come a point where different cleanup is needed, rather than just removing the entry from our database

Reply:
> yeah, I had the same thought; except defer is the thing that always runs at the end, AIUI

From my limited Go experience, you typically set a flag that the callback reads before performing cleanup (i.e: "we've reached the happy path, no need to clean the mess").

Reply:
> there does also come a point where different cleanup is needed, rather than just removing the entry from our database

yeah, this is harder and probably requires splitting this function into smaller ones. I don't think it's worth it for this PR fwiw, just thought of a suggestion :)

Reply:
> > yeah, I had the same thought; except defer is the thing that always runs at the end, AIUI
> 
> From my limited Go experience, you typically set a flag that the callback reads before performing cleanup (i.e: "we've reached the happy path, no need to clean the mess").

I can probably get away with `if err == nil` as the check

Reply:
We use the revert package in lxd to get this conditional behavior that you may find beneficial. 

See https://github.com/lxc/lxd/blob/e1ed3d148f03a0ad32b8975d8dcef76515e4d896/lxd/revert/revert_test.go#L9

Reply:
This also allows to return the reverters Fail function as a revert.Hook to allow a function to return its cleanup steps to the calling function, which can sometimes be useful in more complex call chains.

Reply:
Sorry didn't see @stgraber had said the same thing earlier :)